src/libcore/ptr.rs

```diff
@@ -525,15 +525,41 @@ impl<T: ?Sized> *const T {
         }
     }
 
-    /// Calculates the offset from a pointer. `count` is in units of T; e.g. a
-    /// `count` of 3 represents a pointer offset of `3 * size_of::<T>()` bytes.
+    /// Calculates the offset from a pointer.
+    ///
+    /// `count` is in units of T; e.g. a `count` of 3 represents a pointer
+    /// offset of `3 * size_of::<T>()` bytes.
     ///
     /// # Safety
     ///
-    /// Both the starting and resulting pointer must be either in bounds or one
-    /// byte past the end of an allocated object. If either pointer is out of
-    /// bounds or arithmetic overflow occurs then
-    /// any further use of the returned value will result in undefined behavior.
+    /// If any of the following conditions are violated, the result is Undefined
+    /// Behaviour:
+    ///
+    /// * Both the starting and resulting pointer must be either in bounds or one
+    /// byte past the end of an allocated object.
+    ///
+    /// * The computed offset, **in bytes**, cannot overflow or underflow an
+    ///   `isize`.
+    ///
+    /// * The offset being in bounds cannot rely on "wrapping around" the address
+    ///   space. That is, the infinite-precision sum, **in bytes** must fit in a usize.
```

Comment:
It's possible to construct an offset which wraps around the address space but ends up in bounds, if pointers are modeled as "just" usizes. This would satisfy both previous rules, but fail this one. LLVM explicitly says it's UB to do this with GEP inbounds.